service-registry/src/main/java/org/apache/servicecomb/serviceregistry/consumer/MicroserviceVersions.java

```diff
@@ -191,9 +193,30 @@ private void setInstances(List<MicroserviceInstance> pulledInstances, String rev
         microserviceVersionRule.setInstances(instances);
       }
       revision = rev;
+      // Clean up down versions in case of rollback and other interface not compatible scenarios. Users should ensure running versions are compatible.
+      pack();
     }
   }
 
+  private void pack() {
+    Set<String> microservices = new HashSet<>(instances.size());
+    instances.forEach(item -> {
+      microservices.add(item.getServiceId());
+    });
+    Set<String> downSet = new HashSet<>(instances.size());
+    versions.keySet().forEach(item -> {
+      if (!microservices.contains(item)) {
+        downSet.add(item);
+      }
+    });
+    downSet.forEach(item -> {
+      MicroserviceVersion version = versions.remove(item);
+      for (MicroserviceVersionRule microserviceVersionRule : versionRules.values()) {
+        microserviceVersionRule.deleteMicroserviceVersion(version);
```

Comment:
maybe we just need to update latestVersion?

Reply:
In order to update latestVersion, we also have to check which versions are removed. And I think it is good to remove unused version, because some of them may never used again.

Reply:
yes, but that's not delete from versionRule, but need to delete from MicroserviceVersions